components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java

```diff
@@ -76,23 +78,42 @@ public Response authorize(@Context HttpServletRequest request, MultivaluedMap<St
             return handleErrorResponse(oAuthClientAuthnContext);
         }
 
-        String userCode = GenerateKeys.getKey(Constants.KEY_LENGTH);
-        String deviceCode = UUID.randomUUID().toString();
-        String scopes = request.getParameter(Constants.SCOPE);
-        String redirectionUri = IdentityUtil.getServerURL("/authenticationendpoint/device.do",
-                false, false);
-        String redirectionUriComplete = redirectionUri + "?user_code=" + userCode;
-        deviceAuthService.generateDeviceResponse(deviceCode, userCode, oAuthClientAuthnContext.getClientId(), scopes);
-        return buildResponseObject(deviceCode, userCode, redirectionUri, redirectionUriComplete);
+        try {
+            validateRepeatedParams(request, paramMap);
+            String deviceCode = UUID.randomUUID().toString();
+            String scopes = request.getParameter(Constants.SCOPE);
+            String userCode = getUniqueUserCode(deviceCode, oAuthClientAuthnContext.getClientId(), scopes);
+            String redirectionUri = IdentityUtil.getServerURL(Constants.DEVICE_ENDPOINT_PATH,
```

Comment:
Let's use ServiceURLBuilder utility to build this URL. That will handle cases like tenant qualified paths etc.

A sample can be found here
https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/97d6705182b37beb45482b7fab73351aa9bae29a/components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java#L2635

Reply:
fixed